0030-simple-classical-representations.md

```diff
@@ -0,0 +1,260 @@
+- IBM dynamic-circuits hardware can do bit-level calculations, but there is no way to represent this in Qiskit to pass it on
+- On Terra, we need more of a feel for what will actually help us represent more complex classical constructs
+
+
+## User Benefit
+
+- Users of IBM dynamic circuits will be able to take advantage of the classical-processing capabilities of those compilers from within Qiskit
+- This support is not limited to IBM so may extend further, but we are not aware of other dynamic-circuits hardware accessible through Qiskit
+- For Terra development: this is a "build one to throw away"-type sytem, so we can see in actual usage what works and doesn't for us before we commit to further classical processing.
+
+
+## Design Proposal
+
+The intent of this RFC is to define an exploratory representation for a simple subset of classical bit-level operations, which expand the current state of Terra's `IfElseOp.condition`.
+This representation is not required to be the base for future expansion; it will be permissible to introduce a completely different classical-value type system if that is more appropriate, and deprecate this one.
+To that end, this proposal will not include adding any additional public methods to any existing Terra classes that might conflict with an alternative implementation.
+In particular, this proposal will not consider adding ease-of-use arithmetic methods to `Clbit` or `ClassicalRegister`, nor will it add expressions that can be used as the parameters to gates.
+
+### Representation
+
+`IfElseOp.condition` and `WhileLoopOp.condition` will expand their allowed values to be an "expression" that evaluates to Boolean.
+The allowed expressions will be the following:
+
+- Object:
+  - literal `bool` (`True` or `False`);
+  - literal `uint[_]` (non-negative Python `int`);
+  - runtime `bool` (`Clbit`);
+  - runtime `uint[n]` (`ClassicalRegister(size=n)`).
+- Unary:
+  - `uint[n] -> uint[n]`: `~` (bitwise negation);
+  - `bool -> bool`: `!` (boolean negation).
+- Binary:
+  - `uint[n] * uint[n] -> uint[n]`: `&`, `|`, `^`;
+  - `bool * bool -> bool`: `&&`, `||`;
+  - `uint[n] * uint[n] -> bool`: `==`, `!=`, `<`, `<=`, `>`, `>=`.
+
+For typing in this initial draft, we will make an exact bijection:
+
+- `ClassicalRegister(size=n)` <=> `uint[n]`
+- `Clbit` <=> `bool`
+
+Python Booleans (`True` and `False`) will be used as the `bool` literal values, as they currently are in `Instruction.condition`.
+Non-negative Python ints will be used as the `uint` literal values.
+These do not have an inherent width.
+The width of a literal `uint[_]` will be inferred and fixed to make a containing binary expression validly typed, and mixed-width operations will be forbidden.
+A binary expression in this representation _must_ contain at least one type of fixed width; it will be an error to attempt to use the runtime logic on two literals (this simplifies the implementation, and we won't do constant folding).
```

Comment:
Is there something we can say here about how the width will be inferred? Is this behavior that's coming from OpenQASM3?

Reply:
I'm not sure how to interpret "it will be an error to attempt to use the runtime logic on two literals" in this context. Can you explain a little more?

Reply:
When I wrote this, I was intending `Equal(Value(2), Value(2))` to be a type error on construction, because that expression would involve evaluating a binary op with both inputs having type `uint[UNKNOWN]`, and that's something I wanted to avoid having to define semantics for.

For the inference, this is what I meant:
```python
creg = Value(ClassicalRegister(3))  # type: uint[3]
two = Value(2)  # type: uint[UNKNOWN]
three = Value(3)  # type: uint[UNKNOWN]
five = Value(5)  # type: uint[UNKNOWN]

# This binary op causes the type of `two` to be inferred as `uint[3]` to match.
# This would recurse through any partially type objects.
add_two = Add(creg, two)

# This expression is not valid, because it's a binop on two partial types.
eight = Add(three, five)
```

My intent was just to forbid this kind of thing for now in the actual expression tree, rather than needing to define semantics for how we handle intermediate numeric values of unknown width.  (For example, at what precision would we choose to handle an expression `312890318904839189082032 < 43819432893121004381904`?)

Reply:
Yeah, that would raise or warn becuase `0b1010` can't fit in a `uint[3]`.